rviz_common/include/rviz_common/message_filter_display.hpp

```diff
@@ -151,8 +152,12 @@ class MessageFilterDisplay : public _RosTopicDisplay
 
   virtual void unsubscribe()
   {
-    subscription_.reset();
-    tf_filter_.reset();
+    if (tf_filter_) {
+      tf_filter_.reset();
+    }
+    if (subscription_) {
+      subscription_.reset();
+    }
```

Comment:
I am having a problem with this, as I think it actually shims over the issue here. Calling `reset` on a shared pointer should always work, whether it's pointing to something or is indeed a `nullptr`.

I've just played around with this and I think the real cause is the order in which these two entities are getting destroyed. 
Looking at the initialization, there's `tf_filter_->connectInput(*subscription_);` which couples the `tf_filter` with the `subscription`. I've changed the order here - saying that you call `subscription_.reset()` first and then `tf_filter.reset()` and the issue is gone. So I think that would be the appropriate fix for this here.
But then again, I feel the real problem is actually within the destructor of either entity, the `tf_filter` as well as the message filter to verify that its dependencies don't dangle.

Reply:
First of all, argument `nullptr` is innocent as you pointed :)

However, `tf_filter_->connectInput` is not called on constructor, so `subscription_` has nullptr when this class is initialized.
```cpp
  MessageFilterDisplay()
  : subscription_(nullptr),
    tf_filter_(nullptr),
    messages_received_(0)
  {
    // TODO(Martin-Idel-SI): We need a way to extract the MessageType from the template to set a
    // correct string. Previously was:
    // QString message_type =
    //   QString::fromStdString(message_filters::message_traits::datatype<MessageType>());
    QString message_type = QString::fromStdString("");
    topic_property_->setMessageType(message_type);
    topic_property_->setDescription(message_type + " topic to subscribe to.");
  }
```

`subscription_` gets valid pointer for the first time when `MessageFilterDisplay::subscribe` is called.
So, I think we should check whether `subscription_` is null or not, on the destructor.

Reply:
Sry, I misunderstood and I found that this also works well.
```cpp
  virtual void unsubscribe()
  {
    if (tf_filter_) {
      tf_filter_.message_connection_.disconnect();
    }
    
    subscription_.reset();

    if (tf_filter_) {
      tf_filter_.reset();
    }
  }
```

`tf_filter_.message_connection_` depends on subscription_ via callback function, and this is the underlying cause.
So, we should call `tf_filter_.message_connection_.disconnect` first if `tf_filter_` has valid pointer

Reply:
If you want to reset `subscription_` first, I recommend following code.

```cpp
  virtual void unsubscribe()
  {
    if (tf_filter_) {
      tf_filter_.message_connection_.disconnect();
    }

    if(subscription_) {
      subscription_.reset();
    }

    if (tf_filter_) {
      tf_filter_.reset();
    }
  }
```

`tf_filter_.message_connection_` depends on subscription_ via callback function, and this is the underlying cause.
So, we should call `tf_filter_.message_connection_.disconnect` first if `tf_filter_` has valid pointer

Reply:
Please apologize. I did not realize that your PR already changes the order. What I said earlier is actually wrong and your code is what I had in mind as well. Reset the tf filter, then the subscription. I still do believe we don't need the `if (tf_filter)` and `if(subscription)` clauses.
The following code works for me (additional to the changes you have in the constructor):
```
  virtual void unsubscribe()
  {
    tf_filter_.reset();
    subscription_.reset();
  }
```

Reply:
Thanks for replying.
After I cleaned my workspace and rebuild, I verified that your opinion is right.
I'm sorry too.

I fixed the code :)